Use hourly liquidation events when constructing DLCs

luckysori opened the pull request:
Fix #2299.

The diff is misleadingly massive because when introducing the new feature in https://github.com/itchysats/itchysats/pull/2378/commits/35c4e4e658f8d72f23522a48fa6587cf12222ac5 I chose to simply duplicate the rollover actor. The old version is used to stay compatible with old takers. In my opinion, the correct approach would be to release different versions of `xtra_libp2p_rollover` as we introduce breaking changes. But this is blocked by being able to release some of the crates in our workspace (and `maia`): #2393.

luckysori committed:
Extract rollover protocol into separate crate

The main motivation for this change is to be able to support multiple
versions of the rollover protocol at the same time, without incurring
in lots of duplicated code or increased code complexity due to
branching logic.

Additionally, I think this should allow us to test this protocol in
isolation. I think it's also worth it in terms of the overall
structure of the workspace, and could lead to improved compilation
times.

In my view, the main downside of this approach is that the API of the
rollover protocol is slightly more complicated. We can no longer rely
on knowing the actors and components that we interact with, because
they remain in `daemon`. Instead we depend on 2 new traits:
`model::ExecuteOnCfd` and `rollover::GetAnnouncements`.

The upside of making the rollover actors generic via these traits is
that we can now test this protocol in isolation, as mentioned above.

luckysori committed:
Prefer txid over tx_id

luckysori committed:
Clarify message for IrrelevantAttestation error

luckysori committed:
Model PriceOutOfRange error explicitly

luckysori committed:
Update rollover protocol to use liquidation events

The new `itchysats/rollover/2.0.0` protocol adds liquidation events to
the DLC.

In order to preserve backwards compatibility with takers running
`itchysats/rollover/1.0.0` we run another rollover actor in parallel
for the maker. This leads to a ton of duplicated code, but this
technical debt is temporary and contained in the
`xtra-libp2p-rollover` crate. The idea is to release all our crates to
`crates.io`, so that we can eventually introduce a single cargo
dependency per version of a protocol.

luckysori committed:
Simplify PunishParams

It was not very intuitive to place the identity public keys and the
role in the `xtra_libp2p_rollover::PunishParams` struct. Those values
are not specific to the punishment mechanism of the DLC.

luckysori committed:
Prepare setup_contract::new to use liquidation events

This code used to assume that we had a single event per CFD when
constructing the corresponding DLC. After introducing liquidation
events, that is no longer the case.

Caveat: this code is currently not used! When we added patch
99e62ba6c2d046ec848878c73a15ed2ffed13a40, we chose to only use the
deprecated version of this code (in `setup_contract_deprecated.rs`)
for the maker and all possible takers. The idea was to then use
`setup_contract::new` after adding the contract setup protocol over
the new network layer. Since this hasn't happened at the time of
writing this message, we cannot test the changes in this patch yet.

bonomat commented on model/src/shared_protocol.rs:
```diff
@@ -13,45 +13,40 @@ use maia_core::secp256k1_zkp::XOnlyPublicKey;
 use std::ops::RangeInclusive;
 
-pub(crate) async fn verify_cets(
+pub async fn verify_cets(
```
Looks like it doesn't need async anymore

bonomat commented on xtra-libp2p-rollover/src/maker.rs:
```diff
@@ -137,50 +149,61 @@ where
                 let oracle_pk = self.oracle_pk;
                 let n_payouts = self.n_payouts;
                 async move {
+                    let Rates {
```
I think we will need to add a `ttl` or similar to the rates. 
This is needed in case the automation died/is not functioning. Otherwise we might be rolling over with outdated rates.

luckysori replied:
I suppose you're implying that because we removed the intermediate confirmation step on the maker for rolling over, we don't check that the `automaker` is up before rolling over. And if it is up we understand that the offer must be up-to-date.

I would argue that the current state of affairs of assuming that the offer is up-to-date simply because the `automaker` is up is not much safer than not checking anything at all. Nevertheless, I do think we should introduce a safety mechanism. But I would recommend doing in a separate PR, since this one is quite big already and the situation is not significantly worse with this work.

bonomat replied:
> I suppose you're implying that because we removed the intermediate confirmation step on the maker for rolling over, we don't check that the automaker is up before rolling over. And if it is up we understand that the offer must be up-to-date.

yes, this was what I thought :) 

I think a follow-up PR is fine given the size of this PR. Can you make sure to create a ticket for that once this is merged please?

luckysori replied:
Done: https://github.com/itchysats/itchysats/issues/2531.

bonomat commented on model/src/cfd.rs:
```diff
@@ -2318,21 +2318,26 @@ impl Dlc {
         attestation: &olivia::Attestation,
-    ) -> Result<Result<Transaction, IrrelevantAttestation>> {
-        let cets = match self.cets.get(&attestation.id) {
+    ) -> Result<Result<Result<Transaction, IrrelevantAttestation>, PriceOutOfRange>> {
```
I think I don't see it: why do we need another layer of results?

luckysori replied:
It's not strictly necessary to model it as separate results. It could be modelled as an enum too.

The `PriceOutOfRange` error needs to be handled separately because when we introduce liquidation events we need to ensure that we don't emit an error if the attested price is out of range for a liquididation event (see https://github.com/itchysats/itchysats/commit/151972547b99336f2709b6803cc3345cc211f6f2#diff-f078e5601c18fcb29f773d04d16902354c8f66a57b41aedaea122e11a6f8fd3dR1397-R1402).

bonomat replied:
Ah, now I get it. Thanks.

da-kami replied:
Maybe a solution involving an `enum` would actually be nicer. A tripple `Result` is not particularly beautiful. Not a blocker though.

bonomat commented on model/src/olivia.rs:
```diff
@@ -137,6 +139,32 @@ impl From<Announcement> for maia_core::Announcement {
     }
 }
 
+pub fn hourly_events(
```
What do you think of adding a doc-comment that the resulting list is sorted and adding a test for it?

bonomat replied:
All good, just saw that there are tests for `spaced_events`.

luckysori replied:
I've added some documentation anyway!

bonomat commented on model/src/cfd.rs:
```diff
@@ -2376,6 +2405,23 @@ impl Dlc {
+        self.cets.keys().copied().collect_vec()
+    }
+
+    pub fn liquidation_event_ids(&self) -> Vec<BitMexPriceEventId> {
```
Took me a while, is my understanding correct that a `liquidation_event_id` is every id which is not the `settlement_id`, i.e. the last one? 
It might be worth adding a comment.

luckysori replied:
Correct. I'll add a comment!

bonomat reviewed (commented):
I might have missed it: what's the expectation of how the contract is liquidated? Is the `monitoring` taking care of it?

bonomat commented on model/src/cfd/rollover_v_1_0_0.rs:
```diff
@@ -0,0 +1,156 @@
+use anyhow::Result;
+use time::OffsetDateTime;
+
+impl Cfd {
```
Is anyone running `rollover_v1.0`? 
Afaik we only introduced it with 0.4.19 which is a buggy version and no one is running it. That being said, we could decide to not support `rollover_v1.0` as we don't expect people to use it?

luckysori replied:
Yes, that's a valid point. I considered it, but I didn't know if we knew for certain that no one was running it. And I wanted to take the opportunity to see if the approach with multiple "versions" of the same protocol crate would work.

I don't mind if we decide to delete  the `v_1_0_0` module now, but I don't want to make the call because it's more work for me now haha.

da-kami replied:
Given that we released `0.5.0` with rollover v1 I suppose this comment is irrelevant because we will need both.

bonomat reviewed (approved):
Great work and very clean commits. It's a shame that we can't use it yet because `setup_contract::new` is not used. Do you want to merge this PR anyway or hold back until it is actually being used?

da-kami commented on model/src/shared_protocol.rs:
```diff
@@ -13,45 +13,40 @@ use maia_core::secp256k1_zkp::XOnlyPublicKey;
-                .with_context(|| {
-                    let range = digits.range();
+    for (tx, _, digits) in own_cets.iter() {
+        let counterparty_encsig = counterparty_cets
```
I suppose finding the counterparty encsig within the CETs is significantly less expensive than the verification, so it is fine to move `spwan_blocking` inside. Might be interesting to get some benchmarks at some point.

luckysori replied:
Maybe it's not clear: I've moved it _outside_. The motivation was for the `model` library not to mandate that this computation has to be called from within `spawn_blocking`. And this is only relevant now because we just moved these methods from `daemon` to `model` in https://github.com/itchysats/itchysats/pull/2533/commits/3717fa29f1a78f1f9bc1da6fa95197bbc43362db.

da-kami commented on xtra-libp2p-rollover/src/maker.rs:
```diff
@@ -137,50 +149,61 @@ where
                                     tx_fee_rate,
                                     funding_rate,
-                                    Some(from_params),
+                                    Some((from_event_id, from_complete_fee)),
```
Something in this commit seems odd, `from_event_id` and `complete_fee` are removed, but here we are still using it - where are they coming from? I assume the changes are not cleanly applied after a rebase.

da-kami replied:
My bad

da-kami commented on maker/src/routes.rs:
```diff
@@ -276,6 +276,30 @@ pub async fn get_metrics<'r>(_auth: Authenticated) -> Result<String, HttpApiProb
+
+#[rocket::post("/rollover-config", data = "<config>")]
+#[instrument(name = "POST /rollover-config", skip(maker), err)]
+pub async fn update_rollover_configuration(
```
What is the motivation of this change? Because we remove the `accept` step we need another way to just stop accepting rollovers? What's the use case for this?

da-kami replied:
@bonomat maybe you can explain why this feature is needed. I feel we never really turned rollovers off, so I wonder if this feature is actually justified.

luckysori replied:
@bonomat I'm merging this as is. But we can remove this feature later if it's deemed unnecessary.

bonomat replied:
Sorry, I missed this comment. We might indeed not need this feature.

da-kami commented on maker/src/cfd.rs:
```diff
@@ -589,6 +589,10 @@ impl<O, T, W> Actor<O, T, W> {
         }
     }
 
+    // TODO: Because of this we need to have two rollover actor
```
I think we could actually refactor the old protocol to remove this step as well.
For me the decision if we want to handle `accept` in a different way is independent of the liquidation feature and could be in a separate PR.

luckysori replied:
Note to self: think about why this TODO is still here AND attempt to split this PR into:

- One that removes the accept/reject automation step.
- One that extracts the new crate and introduces the liquidation feature.

luckysori replied:
As promised:

1. https://github.com/itchysats/itchysats/pull/2533 removes the maker's rollover confirmation step.
2. This PR introduces hourly liquidation events.

da-kami commented on model/src/cfd.rs:
```diff
@@ -979,9 +989,11 @@ impl Cfd {
+            rollover::Version::V3 => self.hours_to_extend_in_rollover_based_on_event(
+                *settlement_event_id,
+                now,
+                from_event_id,
```
There is some kind of mixup here, the variable `from_event_id` does not exist.

da-kami replied:
Sorry, my bad

da-kami commented on model/src/payouts.rs:
```diff
@@ -79,3 +117,35 @@ impl Payouts {
+    fn new(announcements: Vec<olivia::Announcement>) -> Result<Self> {
+        let announcements = announcements
+            .into_iter()
+            .sorted_by(|a, b| a.id.cmp(&b.id))
```
Does it make sense to sort using the complete `BitMexPriceEventId` or should we rather sort by the `timestamp` within `BitMexPriceEventId`?

luckysori replied:
Great comment! 

Currently we have derived `PartialOrd` on `BitMexPriceEventId` and the `timestamp` field happens to be the first field, so we are effectively sorting by timestamp. Unless the timestamps are equal, in which case we would compare the `digits`. But those are always `20`, given that we always construct them using `BitMexPriceEventId::with_20_digits`.

But this is a bad approach, since someone could nonchalantly flip the order of the fields and assume that everything is fine. So I will add a patch fixing this.

luckysori replied:
See https://github.com/itchysats/itchysats/pull/2378/commits/46e0890b2c8a105ef63d70c8a2dff2410e903559.

da-kami commented on xtra-libp2p-rollover/src/v_2_0_0/maker.rs:
```diff
@@ -0,0 +1,395 @@
+use xtra_libp2p::Substream;
+use xtra_productivity::xtra_productivity;
+
+/// Permanent actor to handle incoming substreams for the `/itchysats/rollover/1.0.0`
```
```suggestion
/// Permanent actor to handle incoming substreams for the `/itchysats/rollover/2.0.0`
```

da-kami commented on xtra-libp2p-rollover/src/v_2_0_0/maker.rs:
```diff
@@ -0,0 +1,395 @@
+        };
+
+        if !self.is_accepting_rollovers {
+            emit_rejected(order_id, &self.executor).await;
```
So, there is no `accept` anymore, but there is a `reject`?
Does that mean that the taker could either receive a `reject` or the first rollover message in the protocol flow?

da-kami replied:
Sorry, my bad, there is still an accept

da-kami reviewed (commented):
This is great work, thanks for structuring the code into commits.
I think the scope of this PR exploded a bit. 

I feel the decision to remove the rollover confirmation could have done separately to keep this contained. There are so many changes that it becomes hard to evaluate if the complete solution is still sane and no bugs sneaked in. Furthermore, rebasing this work onto the recent rollover changes will not be fun. Maybe it would be worth creating smaller PRs that build on each other next time, so some work can alrady make it into master and we don't have such a big changeset?

I will give this another review after the rebase is done - if you can easily split this work up into multiple PRs that focus on refactor / feature to change `accept` / actual new `rollover_v2` that would help, but I am not sure it's worth the effort at this stage.

luckysori committed:
Remove unused dependencies

Identified using cargo-udeps.

luckysori committed:
Ignore digits when deriving BitMexPriceEventId's PartialOrd

This is important because deriving `PartialOrd` normally means that it
looks at all the struct's fields _in order_. For this type we want to
compare the `timestamp`, but do not care about the `digits`! More
importantly, someone could unintentionally change this behaviour by
reordering the fields.

We also have to ignore the `digits` field for other derived traits,
because inconsistencies there could cause trouble[1].

[1]: https://github.com/rust-lang/rust-clippy/issues/1621.

Co-authored-by: Daniel Karzel <daniel.karzel@coblox.tech>

da-kami commented on CHANGELOG.md:
```diff
@@ -7,6 +7,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+### Added
+
+- Add dynamic liquidation to the DLC, enabling the CFD to be unilaterally closed every hour by either party if the oracle attests to a price close to the ends of the payout curve's domain.
+  This is currently only active _after the first rollover_.
```
Just to make sure my understanding is correct: This means activation with any rollover after rolling this out. If I have a position already open that already had rollovers then it will become active with the first rollover after upgrading.

luckysori replied:
Good question. If you upgrade to version 5.0.0, you will be running `/itchysats/rollover/2.0.0`, so yes.

da-kami commented on model/src/cfd.rs:
```diff
@@ -2350,6 +2379,25 @@ impl Dlc {
+    ///
+    /// This includes:
+    /// - the settlement event; and
+    /// - all the liquidation events that precede it.
```
❓ Would it make sense to separate the settlement event from the liquidation ones? 
Why did you opt for modelling it like this?

luckysori replied:
The main reason for keeping them together is that I don't want to make changes to the `Dlc` because we then have to think about how it affects the database.

When we generate the `Dlc` we, for a very long time, have been collecting the CETs into a `HashMap<BitMexPriceEventId, Vec<Cet>>`. We can rely on the latest `BitMexPriceEventId` (the key of the map) being the settlement event ID. And we already store the settlement event ID separately in the DLC anyway.

In conclusion: we could explicitly model this information in the DLC and I think it would be an improvement. But I don't want to think about breaking changes.

da-kami reviewed (approved):
Just to make sure my understanding is correct: 
With this change we will be able to liquidate hourly (as maker and taker) because we create the necessary liquidation transactions. 
This PR does not include executing the liquidation. I suppose this will be done on top.
